### PR TITLE
Fix batch inference for Whisper model and refactor tests

### PR DESCRIPTION
**Pull request by B-Step62**
### What changes are proposed in this pull request?

Fix the batch inference issue in the Transformers Whisper pipeline.

The Transformers pipeline supports batch inference - if a list of audio data is passed, it will return multiple transcriptions.
```
whiper_pipeline = transformers.pipeline(
    task=task, model=whipsper_model, tokenizer=tokenizer, feature_extractor=feature_extractor
)
whiper_pipeline([audio, audio])
>> [{'text': ' 30-kan vihän konein.'}, {'text': ' 30-kan vihän konein.'}]
```

However, when it is loaded as MLflow Pyfunc model, the model only returns a single transcription and discards the latter.
```
pyfunc_model.predict([audio, audio])
>> ['30-kan vihän konein.']    # <- Only returns single transcription.
```

This PR fixes the issue so that the model returns multiple transcriptions for batch requests, as well as organizing the test cases to increase test coverage while reducing runtime.

### How is this PR tested?

- [x] Existing unit/integration tests
- [x] New unit/integration tests
- [x] Manual tests

Screenshot 2024-07-04 at 19 43 07

Pyfunc model returns multiple transcriptions for a batch request.

(**TODO**: Will do more testing with model serving)

### Does this PR require documentation update?

- [x] No. You can skip the rest of this section.
- [ ] Yes. I've updated:
  - [ ] Examples
  - [ ] API references
  - [ ] Instructions

#### Is this a user-facing change?

- [ ] No. You can skip the rest of this section.
- [x] Yes. Give a description of this change to be included in the release notes for MLflow users.

Fix batch inference issue of the Transformers Whipsper pipeline loaded as MLflow model.

#### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/deployments`: MLflow Deployments client APIs, server, and third-party Deployments integrations
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [x] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>

#### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [x] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

#### Should this PR be included in the next patch release?

`Yes` should be selected for bug fixes, documentation updates, and other small changes. `No` should be selected for new features and larger changes. If you're unsure about the release classification of this PR, leave this unchecked to let the maintainers decide.

<details>
<summary>What is a minor/patch release?</summary>

- Minor release: a release that increments the second part of the version number (e.g., 1.2.0 -> 1.3.0).
  Bug fixes, doc updates and new features usually go into minor releases.
- Patch release: a release that increments the third part of the version number (e.g., 1.2.0 -> 1.2.1).
  Bug fixes and doc updates usually go into patch releases.

</details>

- [x] Yes (this PR will be cherry-picked and included in the next patch release)
- [ ] No (this PR will be included in the next minor release)

### CODE REVIEW

File: mlflow/transformers/__init__.py [hunk 1]
```diff
@@ -2633,94 +2640,65 @@ def process_input_element(input_element):
...
+                   This is the outcome of the pyfunc signature validation for the audio input.
+                   E.g. [{[0]: <audio data>}, {[1]: <audio data>}]
 
-        This utility will convert this JSON encoded, base64 encoded text back into bytes for
-        input into the AutomaticSpeechRecognitionPipeline for inference.
+        Returns:
+            A single or list of audio data.
         """
+        if isinstance(data, list):
+            data = [list(element.values())[0] for element in data]
```
Comment (B-Step62) [inline mlflow/transformers/__init__.py:2660]: note: This is the main change

File: mlflow/transformers/__init__.py [hunk 2]
```diff
@@ -2633,94 +2640,65 @@ def process_input_element(input_element):
...
-        # a pd.DataFrame(pd.Series([raw_bytes])) and then cast to JSON serializable data in the
-        # format:
-        # {[0]: [{[0]: <audio data>}]}
-        # In the inputs format, due to the modification of how types are not enforced, the
-        # logic that is present in processing `records` and `split` format orientation when casting
-        # back to dictionary does not do the automatic decoding of the data from base64 encoded
-        # back to bytes. This is the reason for the conditional logic within `decode_audio` based
-        # on whether the bytes data is base64 encoded or standard bytes format.
-        # The output of the conversion present in the conditional structural validation below is
-        # to return the only input format that the audio transcription pipeline permits:
```
Comment (B-Step62) [inline mlflow/transformers/__init__.py:2710]: note: The pipeline indeed accepts multiple bytes input.

File: mlflow/transformers/signature.py [hunk 3]
```diff
@@ -90,7 +90,7 @@ def infer_or_get_default_signature(
     For signature inference in some Pipelines that support complex input types, an input example
     is needed.
     """
-    if example:
+    if example is not None:
```
Comment (B-Step62) [inline mlflow/transformers/signature.py:93]: note: This is required to allow signature inference with pandas array (the original line raises error of ambiguous boolean casting)

File: tests/helper_functions.py [hunk 4]
```diff
@@ -211,6 +212,19 @@ def pyfunc_serve_and_score_model(
...
+        model_uri,
+        extra_args=extra_args,
+        activity_polling_timeout_seconds=activity_polling_timeout_seconds,
+        stdout=stdout,
+    ) as endpoint:
+        return endpoint.invoke(data, content_type)
+
+
+@contextmanager
+def pyfunc_scoring_endpoint(
```
Comment (B-Step62) [inline tests/helper_functions.py:226]: note: Splitting the server start and inference so we can reuse the same scoring server for multiple requests.

File: tests/transformers/test_transformers_model_export.py [hunk 5]
```diff
@@ -2501,163 +2484,119 @@ def test_instructional_pipeline_with_prompt_in_output(model_path):
...
+    assert pyfunc_transcription == expected_text
 
-    pyfunc_transcription = json.loads(loaded_pyfunc.predict(sound_file_for_test)[0])
-
-    assert transcription["text"] == pyfunc_transcription["text"]
-    # Due to the choice of using tuples within the return type, equivalency validation for the
-    # "chunks" values is not explicitly equivalent since tuples are cast to lists when json
-    # serialized.
-    assert transcription["chunks"][0]["text"] == pyfunc_transcription["chunks"][0]["text"]
+    # 3. Batch prediction with Pyfunc. Float input format is not supported for batch prediction,
```
Comment (B-Step62) [inline tests/transformers/test_transformers_model_export.py:2527]: note: Supporting batch inference for numpy array input might be tricky. Leaving it as unsupported as it's not required for the current customer request.

File: tests/transformers/test_transformers_model_export.py [hunk 6]
```diff
@@ -2501,163 +2484,119 @@ def test_instructional_pipeline_with_prompt_in_output(model_path):
...
+    # because our signature framework doesn't support a list of numpy array.
+    if input_format != "float":
```
Comment (B-Step62) [inline tests/transformers/test_transformers_model_export.py:2529]: note: Supporting batch inference for numpy array input might be tricky. Leaving it as unsupported as it's not required for the current customer request.

**Comment by B-Step62** [comment]
> E   cannot import name 'split_torch_state_dict_into_shards' from 'huggingface_hub' (/opt/hostedtoolcache/Python/3.8.18/x64/lib/python3.8/site-packages/huggingface_hub/__init__.py)

The failed tests with Transformers 4.34.1 seem to be unrelated to this change - also failed in the latest [cross-version test](https://github.com/mlflow-automation/mlflow/actions/runs/9795082576/job/27047843838).

File: tests/helper_functions.py (hunk 4 above)
Comment (harupy) [reply tests/helper_functions.py:226]: Is it possible to land this change first to reduce diff and for easier review?

Comment (B-Step62) [reply tests/helper_functions.py:226]: Sure!

Comment (B-Step62) [reply tests/helper_functions.py:226]: Filed: https://github.com/mlflow/mlflow/pull/12581

Comment (harupy) [reply tests/helper_functions.py:226]: Thanks!

File: tests/transformers/test_transformers_model_export.py [hunk 7]
```diff
@@ -2501,163 +2484,119 @@ def test_instructional_pipeline_with_prompt_in_output(model_path):
...
     )
 
+    # 1. Single prediction with native transformer pipeline
     loaded_pipeline = mlflow.transformers.load_model(model_path)
-
-    transcription = loaded_pipeline(sound_file_for_test, **model_config)
+    transcription = loaded_pipeline(audio)
     assert transcription["text"].startswith(" 30")
+    # strip the leading space
+    expected_text = transcription["text"].strip()
```
Comment (harupy) [inline tests/transformers/test_transformers_model_export.py:2520]:
```suggestion
    # strip the leading space
    expected_text = transcription["text"].lstrip()
```

to match the comment and code

File: tests/transformers/test_transformers_model_export.py [hunk 8]
```diff
@@ -2667,155 +2606,64 @@ def test_whisper_model_serve_and_score_with_timestamps(whisper_pipeline, raw_aud
...
-    artifact_path = "audio_classification"
-    signature = infer_signature(
-        raw_audio_file,
-        mlflow.transformers.generate_signature_output(
-            audio_classification_pipeline, raw_audio_file
-        ),
-    )
+    def _assert_prediction(pred):
+        assert pred["text"] == gt["text"]
+        for pred_chunk, gt_chunk in zip(pred["chunks"], gt["chunks"]):
```
Comment (harupy) [inline tests/transformers/test_transformers_model_export.py:2619]:
```suggestion
        assert len(pred["chunks"]) == len(gt["chunks"])
        for pred_chunk, gt_chunk in zip(pred["chunks"], gt["chunks"]):
```

to ensure `pred["chunks"]` and `gt["chunks"]` have the same length.

**Review by harupy — APPROVED** [review]
left a few comments, otherwise LGTM!

File: tests/transformers/test_transformers_model_export.py (hunk 8 above)
Comment (harupy) [reply tests/transformers/test_transformers_model_export.py:2619]: In Python >= 3.10, the strict argument is available, but we can't use it yet:

https://docs.python.org/3/library/functions.html#zip

File: tests/transformers/test_transformers_model_export.py (hunk 6 above)
Comment (BenWilson2) [reply tests/transformers/test_transformers_model_export.py:2529]: The input signature for a collection of numpy arrays would be a tensor, but the transformers audio implementation doesn't support tensor input. Should we provide an exception message to alert users who might try this to suggest to use a different input format for batch inference of audio transcriptions?

**Review by BenWilson2 — APPROVED** [review]
Great cleanup of the tests! One small food for thought comment about input type handling user notification, but other than that, LGTM!

File: tests/transformers/test_transformers_model_export.py (hunk 8 above)
Comment (B-Step62) [reply tests/transformers/test_transformers_model_export.py:2619]: great catch!

File: tests/transformers/test_transformers_model_export.py (hunk 6 above)
Comment (B-Step62) [reply tests/transformers/test_transformers_model_export.py:2529]: > Should we provide an exception message to alert users who might try this to suggest to use a different input format for batch inference of audio transcriptions?

Hmm the tricky thing is that the exception happens when enforcing schema in pyfunc, before entering to the Transformers wrapper.

> E               mlflow.exceptions.MlflowException: Failed to enforce schema of data '[array([-0.01484766, -0.01701006, -0.01464985, ...,  0.12725656,
E                       0.2055437 ,  0.14151388], dtype=float32), array([-0.01484766, -0.01701006, -0.01464985, ...,  0.12725656,
E                       0.2055437 ,  0.14151388], dtype=float32)]' with schema '[Tensor('float32', (-1,))]'. Error: This model contains a tensor-based model signature with no input names, which suggests a numpy array input or a pandas dataframe input with proper column values, but an input of type <class 'list'> was found.

To translate this exception to the special message, we need to check if transformer model is audio pipeline or not within PythonModel [here](https://github.com/mlflow/mlflow/blob/2bbf9e6d3826a7ea153477b609cfa2a8b513a3a7/mlflow/pyfunc/__init__.py#L699). It's doable, but I'm not very comfortable with leaking flavor logic to general PythonModel implementation.

My opinion is that, given Transformers model itself doesn't support batch inference for float array, there shouldn't be many users who think it's possible in MLflow. Then we don't have to worry too much, at least until getting explicit question from someone who tries that. Alternatively, we could add a quick note in the documentation, but it's might bit challenging to make it easy to reach. What do you think?

Comment (BenWilson2) [reply tests/transformers/test_transformers_model_export.py:2529]: This logic does not belong in pyfunc. Agree with the original design here due to overall complexity and it not really belonging in the parsing logic that has nothing to do with schema enforcement.

**Pull request merged**